### PR TITLE
Update guide and reference of NAS

### PR DESCRIPTION
**Pull request by ultmaster**

### CODE REVIEW

File: docs/en_US/NAS/NasGuide.md [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+.. Note:: The APIs are in an experimental stage. The current programing interface is subject to change.
+```
+
+![](../../img/nas_abstract_illustration.png)
+
+Modern Neural Architecture Search (NAS) methods usually incorporate [three dimensions][1]: search space, search strategy, and performance estimation strategy. Search space often contains a limited neural network architectures to explore, while search strategy samples architectures from search space, gets estimations of their performance, and evolves itself. Ideally, search strategy should find the best architecture in the search space and report it to users. After users obtain such "best architecture", many methods use a "retrain step", which trains the network with the same pipeline as any traditional model.
+
+## Implement a Search Space
+
+Assuming now we've got a baseline model, what should we do to be empowered with NAS? Take [MNIST on PyTorch](https://github.com/pytorch/examples/blob/master/mnist/main.py) as an example, the story looks like this:
```
Comment (QuanluZhang) [inline docs/en_US/NAS/NasGuide.md]: story?

File: docs/en_US/NAS/NasGuide.md [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+        x0 = self.skipcon([x])  # choose one or none from [x]
+        x = self.conv3(x)
+        if x0 is not None:  # skipconnection is open
+            x += x0
+        x = F.max_pool2d(x, 2)
+        # ... same ...
+        return output
+```
+
+The general purpose of InputChoice is a callable module that receives a list of tensors and output the concatenation/sum/mean of some of them, or `None` if none is selected. As a module, we **need to initialize `InputChoice` in `__init__`**, and use it in `forward`. We will see later that this is to allow search algorithms to identify these choices, and do necessary preparation.
```
Comment (QuanluZhang) [inline docs/en_US/NAS/NasGuide.md]: better to tell the default op: concat or sum or mean?

Comment (QuanluZhang) [reply docs/en_US/NAS/NasGuide.md]: "As a module, we **need to initialize `InputChoice` in `__init__`**, and use it in `forward`.", this sentence is not clear, you can either make it clear or add reference to show where you explained this in detail

File: docs/en_US/NAS/NasGuide.md [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+
+Users can specify a **key** for each mutable. By default NNI will assign one for you that is globally unique, but in case users want to share choices (for example, there are two `LayerChoice` with the same candidate operations, and you want them to have the same choice, i.e., if first one chooses the i-th op, the second one also chooses the i-th op), they can give them the same key. The key marks the identity for this choice, and will be used in dumped checkpoint. So if you want to increase the readability of your exported architecture, manually assigning keys to each mutable would be a good idea.
```
Comment (QuanluZhang) [inline docs/en_US/NAS/NasGuide.md]: "for each mutable", seems you have not explained `mutable`

File: docs/en_US/NAS/NasGuide.md [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+
+`LayerChoice` and `InputChoice` are both **mutables**. For advanced usage on mutables, see [Mutables](./NasReference.md#mutables). 
+
+## Use a Search Algorithm
+
+Different in how the search space is explored and trials are spawned, there are at least two different ways users can do search. One runs NAS distributedly, which can be as naive as enumerating all the architectures and training each one from scratch, or leveraging more advanced technique. Since training many different architectures are known to be expensive, another family of methods, called one-shot NAS, builds a supernet containing every candidate in the search space as its subnetwork, and in each step a subnetwork or combination of several subnetworks is trained.
```
Comment (QuanluZhang) [inline docs/en_US/NAS/NasGuide.md]: “leveraging more advanced technique” better to add reference here

File: docs/en_US/NAS/NasGuide.md [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+                           batch_size=128
+                           num_epochs=10,  # 10 epochs
+                           dataset_train=dataset_train,
+                           dataset_valid=dataset_valid,
+                           log_frequency=10)  # print log every 10 steps
+trainer.train()  # training
+trainer.export(file="model_dir/final_architecture.json")  # export the final architecture to file
+```
+
+Users can directly run their training file by `python(3) train.py`, without `nnictl`. After training, users could export the best one of the found models through `trainer.export()`.
```
Comment (QuanluZhang) [inline docs/en_US/NAS/NasGuide.md]: you can simply write "python3 train.py"

File: docs/en_US/NAS/NasGuide.md [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+
+Normally, the trainer exposes a few things that you can customize, for example, loss function, metrics function, optimizer, and datasets. These should satisfy the needs from most usages, and we do our best to make sure our built-in trainers work on as many models, tasks and datasets as possible. But there is no guarantee. For example, some trainers have assumption that the task has to be a classification task; some trainers might have a different definition of "epoch" (e.g., an ENAS epoch = some child steps + some controller steps). So after a few tryouts, if you want to actually use the trainers on your very customized applications, you might very soon need to [customize your trainer](#extend-the-ability-of-one-shot-trainers).
```
Comment (QuanluZhang) [inline docs/en_US/NAS/NasGuide.md]: "exposes a few things" -> "exposes a few arguments"

**Comment by QuanluZhang** [comment]
@ultmaster suggest to add a quick start to show usage, basically it could include how to run a NAS example, how to use users' own dataset, how to export model, how to retrain a model, how to visualize results, etc.

File: docs/en_US/NAS/NasGuide.md [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+model = Net()
+
+# get the chosen architecture from tuner and apply it on model
+get_and_apply_next_architecture(model)
+train(model)  # your code for training the model
+acc = test(model)  # test the trained model
+nni.report_final_result(acc)  # report the performance of the chosen architecture
+```
+
+The search space should be automatically generated and sent to tuner. As with NNI NAS API the search space is embedded in user code, users could use "[nnictl ss_gen](../Tutorial/Nnictl.md)" to generate search space file. Then, put the path of the generated search space in the field `searchSpacePath` of `config.yml`. The other fields in `config.yml` can be filled by referring [this tutorial](../Tutorial/QuickStart.md).
```
Comment (QuanluZhang) [inline docs/en_US/NAS/NasGuide.md]: remove "automatically"

File: docs/en_US/NAS/NasGuide.md [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+
+You could use [NNI tuners](../Tuner/BuiltinTuner.md) to do the search. Currently, only PPO Tuner supports NAS search space.
+
+We support standalone mode for easy debugging, where you could directly run the trial command without launching an NNI experiment. This is for checking whether your trial code can correctly run. The first candidate(s) are chosen for `LayerChoice` and `InputChoice` in this standalone mode.
```
Comment (QuanluZhang) [inline docs/en_US/NAS/NasGuide.md]: I think we should highlight this part, in dedicated section or in quick start.

File: docs/en_US/NAS/NasGuide.md [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+### Retrain with Exported Architecture
+
+After the searching phase, it's time to train the architecture found. Unlike many open-source NAS algorithms who write a whole new model specifically for retraining. We found that searching model and retraining model are usual very similar, and therefore you can construct your final model with the exact model code. For example
+
+```python
+model = Net()
+apply_fixed_architecture(model, "model_dir/final_architecture.json")
+```
+
+After applying, the model is then fixed and ready for a final training. The model works as a single model, although it might contain more parameters than expected. For deeper reasons and possible workaround, see [Trainers](./NasReference.md#retrain).
```
Comment (QuanluZhang) [inline docs/en_US/NAS/NasGuide.md]: suggest to give a link to an example code of retrain

Comment (QuanluZhang) [reply docs/en_US/NAS/NasGuide.md]: also need to explain how to load checkpointed model weights

Comment (QuanluZhang) [reply docs/en_US/NAS/NasGuide.md]: better to explain the format of exported architecture

File: docs/en_US/NAS/NasGuide.md [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+                           dataset_valid=dataset_valid,
+                           log_frequency=10)  # print log every 10 steps
+trainer.train()  # training
+trainer.export(file="model_dir/final_architecture.json")  # export the final architecture to file
+```
+
+Users can directly run their training file by `python(3) train.py`, without `nnictl`. After training, users could export the best one of the found models through `trainer.export()`.
+
+Normally, the trainer exposes a few things that you can customize, for example, loss function, metrics function, optimizer, and datasets. These should satisfy the needs from most usages, and we do our best to make sure our built-in trainers work on as many models, tasks and datasets as possible. But there is no guarantee. For example, some trainers have assumption that the task has to be a classification task; some trainers might have a different definition of "epoch" (e.g., an ENAS epoch = some child steps + some controller steps). So after a few tryouts, if you want to actually use the trainers on your very customized applications, you might very soon need to [customize your trainer](#extend-the-ability-of-one-shot-trainers).
+
```
Comment (QuanluZhang) [inline docs/en_US/NAS/NasGuide.md]: we also should describe distributed training for one-shot nas

File: docs/en_US/NAS/NasGuide.md [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+After the searching phase, it's time to train the architecture found. Unlike many open-source NAS algorithms who write a whole new model specifically for retraining. We found that searching model and retraining model are usual very similar, and therefore you can construct your final model with the exact model code. For example
+
+```python
+model = Net()
+apply_fixed_architecture(model, "model_dir/final_architecture.json")
+```
+
+After applying, the model is then fixed and ready for a final training. The model works as a single model, although it might contain more parameters than expected. For deeper reasons and possible workaround, see [Trainers](./NasReference.md#retrain).
+
+## Customize a Search Algorithm
```
Comment (QuanluZhang) [inline docs/en_US/NAS/NasGuide.md]: suggest to use a dedicated doc file for "customize a search algorithm"

File: docs/en_US/NAS/NasGuide.md [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+
+### Extend the Ability of One-Shot Trainers
+
+Users might want to do multiple things if they are using the trainers on real tasks, for example, distributed training, half-precision training, logging periodically, writing tensorboard, dumping checkpoints and so on. As mentioned previously, some trainers do have support for some of the items listed above; others might not. Generally, there are two recommended ways to add anything you want to an existing trainer: inherit an existing trainer and override, or copy an existing trainer and modify.
+
+Either way, you are walking into the scope of implementing a new trainer. Basically, implementing a one-shot trainer is no different from any traditional deep learning trainer, except that a new concept called mutator will reveal itself. So that the implementation will be different in at least two places:
```
Comment (QuanluZhang) [inline docs/en_US/NAS/NasGuide.md]: you should briefly explain `mutator` here, so that users could get a high level idea of what it is.

Comment (QuanluZhang) [reply docs/en_US/NAS/NasGuide.md]: for example why we introduce `mutator`

File: docs/en_US/NAS/NasGuide.md [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+
+Mutators are designed to be one-to-one to models. One model needs one mutator and one mutator is initialized with one and only one model. Once a mutator is binded to a model, it cannot be rebinded to another model. Mutator is a `nn.Module` so it can have its own parameters -- in fact that's exactly what most algorithms did: they co-optimize parameters in model and parameters in mutator alternatively.
+
+`mutator.reset()` is the core step. That's where all the choices in the model are finalized until next reset. After the reset, the model can be seen as a traditional model to do forward-pass and backward-pass.
+
+Finally, mutators provide a method called `mutator.export()` that export a dict with architectures to the model. Note that currently this dict this a mapping from keys of mutables to tensors of selection. So in order to dump to json, users need to convert the tensors explicitly into python list.
+
+Meanwhile, NNI provides some useful tools so that users can implement trainers more easily. See [Trainers](./NasReference.md#trainers) for details.
+
+### Invent New Mutators
```
Comment (QuanluZhang) [inline docs/en_US/NAS/NasGuide.md]: write a new mutator

File: docs/en_US/NAS/NasGuide.md [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+   
+    .. code-block:: python
+
+        mutator = RandomMutator(model)
+        mutator.reset()
+
+    will immediately set one possible candidate in the search space as the active one.
+```
+
+### Search with Distribution
```
Comment (QuanluZhang) [inline docs/en_US/NAS/NasGuide.md]: we can think about a better title

File: src/sdk/pynni/nni/nas/pytorch/fixed.py [hunk 15]
```diff
@@ -10,20 +10,20 @@
 
 
 class FixedArchitecture(Mutator):
+    """
+    Fixed architecture mutator that always selects a certain graph.
+
+    Parameters
```
Comment (xuehui1991) [inline src/sdk/pynni/nni/nas/pytorch/fixed.py]: same

File: src/sdk/pynni/nni/nas/pytorch/enas/trainer.py [hunk 16]
```diff
@@ -16,64 +16,64 @@
 
 
 class EnasTrainer(Trainer):
+    """
+    ENAS trainer.
+
+    Parameters
```
Comment (xuehui1991) [inline src/sdk/pynni/nni/nas/pytorch/enas/trainer.py]: same

File: src/sdk/pynni/nni/nas/pytorch/enas/mutator.py [hunk 17]
```diff
@@ -28,38 +28,38 @@ def forward(self, inputs, hidden):
 
 
 class EnasMutator(Mutator):
+    """
+    A mutator that mutates the graph with RL.
+
+    Parameters
```
Comment (xuehui1991) [inline src/sdk/pynni/nni/nas/pytorch/enas/mutator.py]: Attributes

File: src/sdk/pynni/nni/nas/pytorch/darts/trainer.py [hunk 18]
```diff
@@ -15,46 +15,46 @@
 
 
 class DartsTrainer(Trainer):
+    """
+    DARTS trainer.
+
+    Parameters
```
Comment (xuehui1991) [inline src/sdk/pynni/nni/nas/pytorch/darts/trainer.py]: Attributes

File: src/sdk/pynni/nni/nas/pytorch/cdarts/trainer.py [hunk 19]
```diff
@@ -32,73 +32,73 @@ def forward(self, student, teacher):
 
 
 class CdartsTrainer(object):
+    """
+    CDARTS trainer.
+
+    Parameters
```
Comment (xuehui1991) [inline src/sdk/pynni/nni/nas/pytorch/cdarts/trainer.py]: Attributes

File: src/sdk/pynni/nni/nas/pytorch/cdarts/mutator.py [hunk 20]
```diff
@@ -127,18 +127,15 @@ def export(self, logger):
 class DartsDiscreteMutator(Mutator):
     """
     A mutator that applies the final sampling result of a parent mutator on another model to train.
+
+    Parameters
```
Comment (xuehui1991) [inline src/sdk/pynni/nni/nas/pytorch/cdarts/mutator.py]: Attributes

File: src/sdk/pynni/nni/nas/pytorch/base_mutator.py [hunk 21]
```diff
@@ -13,7 +13,12 @@
 class BaseMutator(nn.Module):
     """
     A mutator is responsible for mutating a graph by obtaining the search space from the network and implementing
-    callbacks that are called in ``forward`` in Mutables.
+    callbacks that are called in ``forward`` in mutables.
+
+    Parameters
```
Comment (xuehui1991) [inline src/sdk/pynni/nni/nas/pytorch/base_mutator.py]: Attributes

File: docs/en_US/NAS/DARTS.md [hunk 22]
```diff
@@ -43,8 +43,10 @@ python3 retrain.py --arc-checkpoint ./checkpoints/epoch_49.json
...
 
-    .. automethod:: __init__
-
 ..  autoclass:: nni.nas.pytorch.darts.DartsMutator
     :members:
 ```
+
+## Limitations
+
+* DARTS doesn't support DataParallel and needs to be customized in order to support DistributedDataParallel.
```
Comment (scarlett2018) [inline docs/en_US/NAS/DARTS.md]: do we have reference for how to customize the support for distributeddataparallel?

File: docs/en_US/NAS/Overview.md [hunk 23]
```diff
@@ -33,28 +29,32 @@ Here are some common dependencies to run the examples. PyTorch needs to be above
...
 [1]: https://arxiv.org/abs/1802.03268
 [2]: https://arxiv.org/abs/1707.07012
 [3]: https://arxiv.org/abs/1806.09055
 [4]: https://arxiv.org/abs/1806.10282
 [5]: https://arxiv.org/abs/1703.01041
 
-## **Reference and Feedback**
 * To [report a bug](https://github.com/microsoft/nni/issues/new?template=bug-report.md) for this feature in GitHub;
-* To [file a feature or improvement request](https://github.com/microsoft/nni/issues/new?template=enhancement.md) for this feature in GitHub;
-* To know more about [Feature Engineering with NNI](https://github.com/microsoft/nni/blob/master/docs/en_US/FeatureEngineering/Overview.md);
```
Comment (scarlett2018) [inline docs/en_US/NAS/Overview.md]: let's keep the references for FE, Model compression and HPO. That's just for users who would like to explore more features from NNI. no hurt for putting the links there.

File: docs/en_US/NAS/NasGuide.md [hunk 24]
```diff
@@ -0,0 +1,311 @@
...
+
+![](../../img/nas_abstract_illustration.png)
+
+Modern Neural Architecture Search (NAS) methods usually incorporate [three dimensions][1]: search space, search strategy, and performance estimation strategy. Search space often contains a limited neural network architectures to explore, while search strategy samples architectures from search space, gets estimations of their performance, and evolves itself. Ideally, search strategy should find the best architecture in the search space and report it to users. After users obtain such "best architecture", many methods use a "retrain step", which trains the network with the same pipeline as any traditional model.
+
+## Implement a Search Space
+
+Assuming now we've got a baseline model, what should we do to be empowered with NAS? Take [MNIST on PyTorch](https://github.com/pytorch/examples/blob/master/mnist/main.py) as an example, the code might look like this:
+
+```python
```
Comment (scarlett2018) [inline docs/en_US/NAS/NasGuide.md]: after implementing codes for search space, do we need to talk about the SearchSpace json file?
